### ISSUE TITLE
Implementation New closing "method" 

### ISSUE THREAD

**Issue by diogovk** [mathall/rim#6]
In vim we have three commands for "exiting":

:qa - quits the program (Cmd::Quit)
:clo - close current window, if it isn't the last (Cmd::CloseWindow)
:q - close current window, and quits program if it's the last (Not yet implemented?)

I'd like to offer to implement the ":q" in rim. (I'd be starting with something easy)
What do you think?

**Comment by mathall** [comment]
Sorry I didn't manage to answer before your pull request. In general though, thumbs up, go ahead.

I was holding off on this because I wanted the command line to utilize some scripting engine. Of course that's a huge task and there's other ways to go about it. On the other end of the spectrum you could just bind `:qa<Enter>` to Cmd::Quit, (which I now see you've done) and that's one solution I wouldn't be very interested in. `:` bears a more specific meaning, namely you "enter" the command line, and simply binding keys is not solving the problem.

If you think you're up for adding a sort of command line, I think it would be good if you would suggest some rough course of action first. How should the command line behave? What would be the mechanics of interpreting the line fed to it and taking action?

As for your pull request, I wasn't aware of this command. Looks good, but I think the only valid binding is `<C-w>q`, no? Do you mind if I fixup your commit myself before merging?

Always feel free to shoot me an email if you rather discuss anything in private.

**Comment by diogovk** [comment]
> Always feel free to shoot me an email if you rather discuss anything in private.

Here is fine.
Please let me know if there's some other channel where you'd rather discuss the development (irc, group, etc).

> On the other end of the spectrum you could just bind :qa<Enter> to Cmd::Quit

I'm aware that that's far from perfect, but this "suggestion" came from the "frustration" of not knowing how to quit rim the first time I loaded (as a "temporary solution"). As a user of vim I find that plain "q" means start macro recording. 
I'm still not an experienced rust programmer so I though I'd start with something easy.

> wanted the command line to utilize some scripting engine

I'd love to contribute in the script engine, but I feel that's something I'm not ready to tackle alone. To be frank I don't even know where to start.
One thing that I would add (and that's just my opinion) is that we could prioritize the features that will make this software usable in the our "daily lives" (quitting, saving, opening file, arguments parsing). You probably know that pico and nano are quite popular even though they don't have a script engine (although I'm pretty sure they're not our "target user").
By "daily lives" I mean plain-text and config files editing.

> Do you mind if I fixup your commit myself before merging?

I don't mind at all. I also don't mind editing them according to your instructions.
Whichever is more convenient to you!

> the only valid binding is `<C-w>q`, no?

I'm not sure I understand what you mean but in my vim:
:help CTRL-W_CTRL-Q
brings me the :quit command

**Comment by mathall** [comment]
> I'm aware that that's far from perfect, but this "suggestion" came from the "frustration" of not knowing how to quit rim the first time I loaded (as a "temporary solution"). As a user of vim I find that plain "q" means start macro recording.

That's a very fair point. The `q` binding was never meant to stay anyway, lets just remove it already and replace it with `:qa<Enter>` similar to how you've introduced `:q<Enter>`. The same sort of remapping should be done for `<C-w><C-w>` to `:w<Enter>` really. Since these bindings are for convenience until a real solution comes along though, I'd prefer if we don't go overboard with them. And now I see all of these should be in default_mode(), why did I put `<C-w><C-w>` in default_normal_mode()? Welp, lots of small stuff. :P For the pull request, could you just move the new bindings to the correct function? (default_mode) Then I could handle the rest when merging if you haven't already.

> One thing that I would add (and that's just my opinion) is that we could prioritize the features that will make this software usable in the our "daily lives" (quitting, saving, opening file, arguments parsing). You probably know that pico and nano are quite popular even though they don't have a script engine.

Althought this has never been a goal, (reaching "usable" status quickly) I got to say that sounds neat and I'm all for it at this point.

> I'm not sure I understand what you mean but in my vim:
> :help CTRL-W_CTRL-Q
> brings me the :quit command

Oh you're right, my bad. I was just testing in vim and it didn't work. In gvim (or macvim rather) it does work. Maybe it's just Terminal messing with me? I don't know, interesting.

**Comment by diogovk** [comment]
> And now I see all of these should be in default_mode(), why did I put `<C-w><C-w>` in default_normal_mode

To tell the truth, I didn't even think about it, but I went to try in vim `<C-w>q` in insert mode it didn't work in my vim version. `<C-w>` deletes one word backwards.

If you still want me to move I'll move, no problem.

**Comment by mathall** [comment]
You don't need to move that one, it's a mistake I made which I just realised now. But the newly added bindings should go in default_mode(). These won't be handled in insert mode because insert mode will chew up the individual keys before this lower prioritised mode has a chance to string them together. If you're interested the comments in commands.rs explain mode priorities and such fairly well I hope.
